tc/core/tensor-inl.h

```diff
@@ -0,0 +1,322 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <algorithm>
+#include <sstream>
+
+#include <glog/logging.h>
+
+namespace tc {
+
+template <typename DLTensorType>
+inline void setSizes(DLTensorType& t, const std::vector<int64_t>& sizes) {
+  auto ndim = sizes.size();
+  for (size_t i = 0; i < ndim; ++i) {
+    t.shape[i] = sizes[i];
+  }
+}
+
+template <typename DLTensorType>
+inline void setStrides(DLTensorType& t, const std::vector<int64_t>& strides) {
+  auto ndim = strides.size();
+  for (size_t i = 0; i < ndim; ++i) {
+    t.strides[i] = strides[i];
+  }
+}
+
+template <typename T>
+std::vector<T> makeStridesFromSizes(const std::vector<T>& sizes) {
+  auto ndim = sizes.size();
+  if (ndim == 0) {
+    return std::vector<T>();
+  }
+  std::vector<int64_t> strides(sizes.size(), 0);
+  strides[ndim - 1] = 1;
+  for (int i = ndim - 2; i >= 0; --i) {
+    strides[i] = strides[i + 1] * sizes[i + 1];
+  }
+  return strides;
+}
+
+inline DLContext getCPUDLContext() {
+  DLContext res;
+  res.device_id = 0;
+  res.device_type = DLDeviceType::kDLCPU;
+  return res;
+}
+
+inline DLContext getGPUDLContext(int device_id) {
+  DLContext res;
+  res.device_id = device_id;
+  res.device_type = DLDeviceType::kDLGPU;
+  return res;
+}
+
+template <typename T>
+DLDataType getDLDataType();
+
+template <>
+inline DLDataType getDLDataType<float>() {
+  DLDataType res;
+  res.code = DLDataTypeCode::kDLFloat;
+  res.bits = 32;
+  res.lanes = 1;
+  return res;
+}
+
+template <>
+inline DLDataType getDLDataType<int>() {
+  DLDataType res;
+  res.code = DLDataTypeCode::kDLInt;
+  res.bits = 32;
+  res.lanes = 1;
+  return res;
+}
+
+template <typename DLTensorType>
+void DLTensorDeleter::operator()(const DLTensorType* t) {
+  if (t->shape) {
+    delete[] t->shape;
+  }
+  if (t->strides) {
+    delete[] t->strides;
+  }
+  delete t;
+};
+
+template <typename T, typename DLTensorType, typename VoidPtrType>
+inline void fillDLTensor(
+    DLTensorType* res,
+    DLContext ctx,
+    DLDataType dtype,
+    const std::vector<T>& sizes,
+    const std::vector<T>& strides = std::vector<T>(),
+    VoidPtrType* data = nullptr,
```

Comment:
I wouldn't call this `VoidPtrType`, it's just the type of data stored.  Actually, there's no need to instantiate this template for different sizes because the underlying dltensor uses an actual `void*`.  This will only bloat the code, slowing down compilation, with no additional type safety.

Reply:
We have both ```void*``` and ```const void*``` for DLTensor and ```DLConstTensor``` (coming from inputs and outputs of networks as per caffe2 so there is additional type safety).